experimental/dds/tree2/src/feature-libraries/object-forest/objectForest.ts

```diff
@@ -89,8 +91,13 @@ class ObjectForest extends SimpleDependee implements IEditableForest {
 		this.anchors.forget(anchor);
 	}
 
-	public applyDelta(delta: Delta.Root): void {
-		this.events.emit("beforeDelta", delta);
+	public acquireVisitor(): DeltaVisitor {
+		assert(
+			this.activeVisitor === undefined,
+			"Must release existing visitor before acquiring another",
+		);
+		this.events.emit("beforeChange");
+
 		this.invalidateDependents();
 		assert(
```

Comment:
previously this used to be enough to assert that no cursors (other than the one used foe the update) exist during delta application. With this change, this check (and likely a similar one in chunked forest) is no longer enough.

There are two approaches we can take to close this safety gap:
1. Add `assert(this.forest.activeVisitor === undefined)` into Cursor.setToAboveDetachedSequences (this will ban any use of cursors while you have a visitor)
2. OR: in the visitor implementation, before it makes each change, do `this.invalidateDependents(); this.currentCursors.size === 1`, and remove those checks from here (this will allow using cursors inbetween edits, for example inspecting the editiable tree on the delta events.)

Equivalent changes will be needed for chunked forest.

Reply:
If you go with the second option, you might want to move before and after change into each individual edit as well, or at least update the docs on those events to clarify when they fire.

Reply:
Nice catch! Thanks for noticing that. I'm guessing we want option #2 given where we're headed.